### ISSUE TITLE
Two curious questions

### ISSUE THREAD

**Issue by fgergvdsvgsdh** [MaartenGr/BERTopic#275]
1. I want to know why when I run the BerTopic different times I get different results (topics etc..). I am also interested on the theoretical point of view I guess it has something to do with random processes but i don't have a clear view.
2. Once I save the topic model what can I extract from the saved model? Can I extract everything and continue from the loading point like I previously ran the model?

**Comment by cb-pratibhasaha** [comment]
I was curious to know if there a way to make sure that everytime I run Bertopic, I get the same topics/results? 

Thanks in advance.

**Comment by MaartenGr** [comment]
@cb-pratibhasaha 

> I was curious to know if there a way to make sure that everytime I run Bertopic, I get the same topics/results?

The underlying dimensionality reduction algorithm, UMAP, is stochastic which results in different results each run. You can find more about that [here](https://maartengr.github.io/BERTopic/faq.html#why-are-the-results-not-consistent-between-runs) and how to get the same results each time. 

> Also, if it produces different results each time it is run,with the count of different topics differing each time, how do we know it is giving accurate results?

That depends on your definition of "accurate". Seeing as topic modeling can be quite subjective, it really depends on your use case, your evaluation metrics, the stakeholders involved etc. From an algorithmic perspective, you can still set a `random_state` in UMAP which is similar to what you can do in many other algorithms.

**Comment by cb-pratibhasaha** [comment]
Thank you for the response @MaartenGr.  I used the following code and set the random state to specific number yet it gives a different result every time I run :

from bertopic import BERTopic
from sklearn.cluster import KMeans
from umap import UMAP
from hdbscan import HDBSCAN
cluster_model = KMeans(n_clusters=150)
umap_model = UMAP(random_state=42)
topic_model = BERTopic(hdbscan_model=cluster_model, vectorizer_model = vectorizer_model_cb, umap_model = umap_model)
topics ,_ = topic_model.fit_transform(text) 

Do let me know your thoughts. Again, thank you for the response.

**Comment by cb-pratibhasaha** [comment]
As an extension, for choosing the clustering algorithm, 
HBDSCAN gives too many outliers while K-Means pushes outliers into categories they do not belong to. I wanted to ask if there is a clustering algorithm that is sort of a middle ground to both these algorithms such that the proportion of outliers is reduced without compromising accuracy?

**Comment by MaartenGr** [comment]
> Thank you for the response @MaartenGr. I used the following code and set the random state to specific number yet it gives a different result every time I run :

This is because `KMeans` also has a `random_state` parameter that you should use that make sure every run is the same. 

> As an extension, for choosing the clustering algorithm,
HBDSCAN gives too many outliers while K-Means pushes outliers into categories they do not belong to. I wanted to ask if there is a clustering algorithm that is sort of a middle ground to both these algorithms such that the proportion of outliers is reduced without compromising accuracy?

I would advise playing around with some of the `HDBSCAN`  parameters. Especially lowering `min_samples` and `min_cluster_size` can result in fewer outliers. You can find more about that [here](https://maartengr.github.io/BERTopic/faq.html#how-do-i-reduce-topic-outliers). That way, you have some control over which portion of outliers suits your use case.

**Comment by MaartenGr** [comment]
@Kuniko925 Could you share your entire code or create a reproducible example? Without it, it is quite difficult to understand what is happening in your specific environment. Also, could you share the version of BERTopic and its dependencies?

**Comment by Kuniko925** [comment]
@MaartenGr 
Thank you very much for your reply. I am sorry that I did not share my code.
Version of bertopic  is 0.14.1.

`
from bertopic import BERTopic
from sentence_transformers import SentenceTransformer, util
from umap import UMAP
from hdbscan import HDBSCAN
from sklearn.feature_extraction.text import CountVectorizer
from bertopic.vectorizers import ClassTfidfTransformer
from sklearn.cluster import KMeans

import nltk
from nltk.corpus import stopwords
nltk.download("punkt")
nltk.download("wordnet")
nltk.download("omw-1.4")
nltk.download("english")
nltk.download("stopwords")

# Reference URL: https://github.com/MaartenGr/BERTopic/issues/286
from nltk import word_tokenize          
from nltk.stem import WordNetLemmatizer

class LemmaTokenizer:
  def __init__(self):
    self.wnl = WordNetLemmatizer()
  def __call__(self, doc):
    # [abbriviation_converter(self.wnl.lemmatize(t)) for t in word_tokenize(doc)]
    return [self.wnl.lemmatize(t) for t in word_tokenize(doc)]

def modeling_bertopic():

  n_neighbors = 30
  n_components = 100
  min_cluster_size = 25
  top_n_words = 30
  min_samples = 1
  ngram_range = (1, 3)

  sentence_model = SentenceTransformer("all-MiniLM-L12-v2")
  umap_model = UMAP(n_neighbors=n_neighbors, n_components=n_components, min_dist=0.0, metric="cosine", random_state=42)
  hdbscan_model = HDBSCAN(min_cluster_size=min_cluster_size, metric="euclidean", cluster_selection_method="eom", prediction_data=True, min_samples=min_samples)
  # cluster_model = KMeans(n_clusters=20)
  ctfidf_model = ClassTfidfTransformer(reduce_frequent_words=True, bm25_weighting=True)
  vectorizer_model = CountVectorizer(ngram_range=ngram_range, max_df=0.70, tokenizer=LemmaTokenizer(), stop_words=stopwords.words("english"))

  model = BERTopic(nr_topics = "auto", language="english", top_n_words=top_n_words,
                       embedding_model=sentence_model, umap_model=umap_model, hdbscan_model=hdbscan_model, 
                       ctfidf_model=ctfidf_model, vectorizer_model=vectorizer_model,
                      calculate_probabilities=True)
  
  return model

topic_model = modeling_bertopic()
topics, probs = topic_model.fit_transform(abstract)
`

Thank you very much again.

**Comment by MaartenGr** [comment]
@Kuniko925 Could you also try it with k-Means instead of HDBSCAN? It might be that there is something going on there. Also, perhaps not setting `nr_topics` at all might be worthwhile to try out. Other than that, I believe it should create reproducible results. Iteratively removing a parameter that each runs to see where the main issue lies might help. 

NOTE: I edited your message to make sure the code is displayed in a readable format.

**Comment by MaartenGr** [comment]
@Kuniko925 No problem at all! Glad to hear that you fixed the issue.

**Comment by alicjamalota** [comment]
> @MaartenGr Sorry to bother you repeatedly. I tried removing a parameter of UMAP, n_components. Then so far, the same topics have been reproduced. Thank you very much for your support.

Hey! What did you mean by 'removing' the n_components parameter? Thanks!

**Comment by alicjamalota** [comment]
Hey @MaartenGr ! 

I am facing the same problem as discussed above. I have different results with each run of BERTopic even though I set the random_state=42 for UMAP and I do not use any HDBSCAN (it is by default set to None). I really want to find a solution for that problem. I will appreciate your help on that. Thanks!

main_representation = KeyBERTInspired()
aspect_model2 = [KeyBERTInspired(top_n_words=30), MaximalMarginalRelevance(diversity=.5)]
aspect_model3 = PartOfSpeech()

representation_model = {
   "Main": main_representation,
   "Representation2":  aspect_model2,
   "Representation3":  aspect_model3 
}

topic_model = BERTopic(verbose=True,
                       umap_model=UMAP(n_neighbors=15,n_components=5, min_dist=0.0, metric='cosine', random_state=42), 
                       seed_topic_list=seed_topic_list, 
                       vectorizer_model=CountVectorizer(stop_words=stop_words, max_df=1.0, ngram_range=(1,2)),
                       representation_model=representation_model 
                      ).fit(df.preprocessed_text, y=df.ssot_topic_id)
